### ISSUE TITLE
Filter events by tag and its base ref

### ISSUE THREAD

**Issue by msuterski** [drone/drone#2217]
Our use case requires that we would only run certain builds on tag events based on a specific branch. 

Specifically, it's to ensure that the tag event triggers a prod deployment only from master, which is a protected branch with restricted write access. 

We're using github, so my references will be limited to it.

I couldn't find a way to do it in the drone's yml syntax or a reference to `base_ref` in the drone's source. 

I'd be happy to file a PR with this change. I could either add a new option to `events` (`base_branch` ?) or override the `Build.Ref` with a value of `base_ref` in case the event is for a tag.

**Comment by bradrydzewski** [comment]
ok yes, I see what you mean. I see the `base_ref` field in the payload

```js
{
  "ref": "refs/tags/v0.8.0-rc.5",
  "before": "0000000000000000000000000000000000000000",
  "after": "538f25683450489a96808bc11cb5400c1939ac82",
  "created": true,
  "deleted": false,
  "forced": false,
  "base_ref": "refs/heads/master",
  "compare": "https://github.com/drone/drone/compare/v0.8.0-rc.5",
  "commits": [

  ]
  /* truncated */
}
```

I think we can safely set the `build.Branch` to the `base_ref` for tag events. I do not think we need a new field here, and I don't think there will be any adverse effects using the branch field.

You will then be able to do something like this:

```
when:
  event: tag
  branch: master
```

Note that we probably want to implement the following basic logic just to be safe:

* only use the `base_ref` if starts with `refs/heads/`
* when setting `base_ref` to the branch, we should trim prefix `refs/heads/`

I would definitely accept a PR for this as described above, if you think it would solve your use case.

**Comment by msuterski** [comment]
Great. I'll start working it. Thanks.

**Comment by tboerger** [comment]
But branch is in case of a tag already `refs/tags/v0.8.0-rc.5`?

**Comment by tboerger** [comment]
I'm already using something like this:

```
when:
  event: [push, tag]
  branch: [master, refs/tags/*]
```

**Comment by bradrydzewski** [comment]
@tboerger I just added ref as a proper constraint:
https://github.com/cncd/pipeline/commit/b7427ae1fb86bb04e2997e5f3a1eec75c4b056ef

so we can do something like this:

```
when:
  event: [ push, tag ]
  ref: [ refs/heads/master, /refs/tags/* ]
```

I think this makes a bit more sense, since storing the ref in the branch only works in github. This approach should work with all providers.

**Comment by bradrydzewski** [comment]
@msuterski would you mind running the below commands to fetch the changes I made to one of our vendored dependencies? Thanks!

```
go get -u github.com/cncd/pipeline/...
govendor update github.com/cncd/pipeline/pipeline/...
``

**Comment by msuterski** [comment]
sure.

**Comment by msuterski** [comment]
I'm not familiar with drone's internal, but if we're going to reuse the `Branch` and override it with `base_ref`, won't that be a breaking change? 

I'm thinking of a case where someone could have their builds triggered on a `tag` and also included the `branch: ref/tags/*`? Like what @tboerger had but more specific, like run on tags with prefix in a name:
```
when:
  event: tag
  branch: refs/tags/prod-deploy-*
```

**Comment by bradrydzewski** [comment]
It is true this would be a breaking change and we would have to document accordingly in the release notes so that everyone is aware.

I would describe the current behavior as more of an unintended quirk. The only provider that uses the tag ref as the branch value is GitHub and this wasn't really intentional, despite it being used today as a workaround.

I think this is therefore more of a permanent fix and something I approve breaking before 1.0. So now would be the right time for this sort of breaking change.

**Comment by tboerger** [comment]
Ref sounds like a proper solution

**Comment by msuterski** [comment]
👍 sounds good to me.

**Comment by bradrydzewski** [comment]
closed with #2218 (thanks !)

**Comment by anthosz** [comment]
Hello,

Do you know how to use tag + branch for tags & branch demo?

I try these conditions without success:
    when:
      event: [ push, tag ]
      branch: [ demo, refs/tags/* ]
    when:
      event: [ push, tag ]
      branch: [ demo ]
      ref: refs/tags/*
    when:
      event: [ push, tag ]
      branch: [ demo, refs/tags/* ]
      ref: refs/tags/*

The issue is that I don't want to execute this step on master branch but only for tag & demo branch.
